katdal/applycal.py

```diff
@@ -22,41 +22,66 @@
 import numpy as np
 import dask.array as da
 
-from katdal.categorical import CategoricalData
+from .categorical import CategoricalData, ComparableArrayWrapper
 
 
-def calc_delay_correction(cache, product, index, freqs):
-    """Calculate correction sensor from delay calibration solutions.
+def get_cal_product(cache, attrs, product):
+    """Extract calibration solution `product` from `cache` as a sensor.
 
-    This extracts the cal solution sensor associated with `product` from
-    `cache`, then extracts the delay time series of the input specified by
-    `index` (in the form (pol, ant)) and builds a categorical sensor for
-    the corresponding complex correction terms (channelised by `freqs`).
+    This takes care of stitching together multiple parts of the product
+    if this is indicated in the `attrs` dict.
+    """
+    key = 'cal_product_' + product
+    try:
+        parts = int(attrs[key + '_parts'])
+    except (KeyError, TypeError):
```

Comment:
When would you get TypeError?

Reply:
You should ask yourself - I copied that code from `LoaderKatdal._load_cal_product` :-)

I suspect you wanted to ensure that `cal_product_X_parts` is an int, but why would it not be? I'm in favour of just doing `KeyError` here if you are.

Reply:
Done.